VMs: metrics: add scripts to generate metrics CI VMs

grahamwhaley opened the pull request:
Add a set of scripts to configure, generate and clone virsh
VMs for use under a metrics Jenkins CI system.

Fixes: #29

Signed-off-by: Graham whaley <graham.whaley@intel.com>

jodh-intel commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+ 	-n ${VM_NAME} \
+ 	--description "simple test" \
+ 	--os-type=Linux \
+ 	--os-variant=ubuntu16.04 \
```
It might be an idea to define this as a variable at the top to make it stand out and be easier to maintain.

grahamwhaley replied:
actually, I think that is a bug! that os-variant should probably be a VAR derived from each specific image subdir... an oversight. I'll fix that.

jodh-intel commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+ 	--cpu=host \
+ 	--ram=8192 \
+ 	--vcpus=2 \
+ 	--disk path=${IMG_NAME},bus=virtio,size=10 \
```
I'd use a variable for both this `10` and the `10G` arg to `qemu-img` above.

jodh-intel commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+	die
+fi
+
+CONFIG_DIR=$1
```
I'd double-quote here (and in the other scripts) to handle spaces in directory names.

jodh-intel commented on VMs/metrics/3_complete_baseimage.sh:
```diff
@@ -0,0 +1,59 @@
+
+source ${CONFIG_DIR}/${ENV_FILE}
+
+# We could be smarter and check if it is already running...
```
This looks like a FIXME? If so, you could turn it into an issue and reference here.

grahamwhaley replied:
I'll make it a FIXME. There are a bunch of extra smarts that could be added to the scripts, but prudency to get it working and out took priority

jodh-intel commented on VMs/metrics/3_complete_baseimage.sh:
```diff
@@ -0,0 +1,59 @@
+echo "IP of ${VM_NAME} is ${VM_IP}"
+
+message "Checking VM uname"
+VM_UNAME=$(ssh -F /dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -i ${PRIVKEYNAME} ${VM_USERNAME}@${VM_IP} uname -n)
```
These `ssh`/`scp` options are used in a few places so I'd put them in a variable and reference that to reduce line length and make it easier to maintain those options.

jodh-intel commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+
+source global_env.sh
+
+die() {
```
These functions are defined in all the scripts, so how about putting them into a `lib.sh` and `source`-ing?

grahamwhaley replied:
yep, did occur to me - will do

jodh-intel commented on VMs/metrics/6_start_clone_agent.sh:
```diff
@@ -0,0 +1,76 @@
+
+	# And make the clone
+	message "Cloning ${VM_NAME} -> ${VM_CLONE_NAME}"
+	./4_clone_vm.sh ${CONFIG_DIR}
```
I was assuming these scripts had a numeric prefix so they were run in that order. But if so, why is `7` run before `4`?

jodh-intel commented on VMs/metrics/6_start_clone_agent.sh:
```diff
@@ -0,0 +1,76 @@
+
+CONFIG_DIR=$1
+
+# We are not allowed to generate any output or consume anything from stding or
```
Typo: "stdin".

jodh-intel commented on VMs/metrics/global_env.sh:
```diff
@@ -0,0 +1,18 @@
+USER_DATA_NAME=user-data
+
+VM_USERNAME=jenkins
+VM_BOOTTIME=30
```
Presumably this is in seconds?

jodh-intel commented on VMs/metrics/ubuntu16.04/env.sh:
```diff
@@ -0,0 +1,12 @@
+PRIVKEYNAME=$(pwd)/virsh.rsa
+PUBKEYNAME=${PRIVKEYNAME}.pub
+
+VM_NAME=ubnt16.04_master
```
Is this a typo ("ubuntu"), or just truncated to fit into some limit?

grahamwhaley replied:
I think this name trimming is related to the libvirt XML related name rules https://libvirt.org/formatdomain.html#elementsMetadata
I'll see if I can come up with a fuller name that fits both the libvirt name and the subsequent uname, captures the distro better and allows us to distinguish between Master and Clone images...

jodh-intel commented on VMs/metrics/ubuntu16.04/env.sh:
```diff
@@ -0,0 +1,12 @@
+PUBKEYNAME=${PRIVKEYNAME}.pub
+
+VM_NAME=ubnt16.04_master
+VM_CLONE_NAME=ubnt16.04_clone
```
Same comment.

jodh-intel commented on VMs/metrics/ubuntu16.04/env.sh:
```diff
@@ -0,0 +1,12 @@
+
+VM_NAME=ubnt16.04_master
+VM_CLONE_NAME=ubnt16.04_clone
+IMG_BASE=${IMG_DIR}/xenial-server-cloudimg-amd64-disk1.img
```
Hard-coded architecture. Could you change this to use `$(dpkg --print-architecture)` as this file is Ubuntu-specific?

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data:
```diff
@@ -0,0 +1,54 @@
+      - PUBLICKEYCONTENTS
+
+packages:
+ - gawk
```
Nit: You could sort this list for bonus points ;)

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data:
```diff
@@ -0,0 +1,54 @@
+
+ - rm /etc/update-motd.d/10-help-text /etc/update-motd.d/51-cloudguest
+ - rm /etc/update-motd.d/90-updates-available
+ - rm /etc/legal
```
OOI why?

grahamwhaley replied:
I inherited it from https://github.com/intel/ccloudvm/blob/master/workloads/clearcontainers-3.x.yaml#L46
I suspect it has to do with the rules for modifying and 'shipping' an Ubuntu image.
@markdryan @mcastelino - are you able to clarify for us?

markdryan replied:
This came from the original ccloudvm workload, ciao.  This workload displays a prompt every time you log into the VM explaining how to start running the ciao development environment.  These three files where removed, as their presence caused the ciao start up instructions to get lost in a whole pile of other information.

grahamwhaley replied:
Thanks @markdryan - in which case I think i can drop those lines.

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data:
```diff
@@ -0,0 +1,54 @@
+
+# Install Go
+# The sourceing of environment is a workaround to get any proxy settings we need
+ - 'bash -c "set -a; source /etc/environment; set +a; curl -L https://dl.google.com/go/go1.10.2.linux-amd64.tar.gz -o /tmp/go1.10.2.linux-amd64.tar.gz"'
```
Hard-coded golang version and architecture in this block.

grahamwhaley replied:
I know. We only need it for the initial 'go get' of the test repos, and then the CI scripts install a new version...
Longer term maybe I will swap it out for a 'git clone'. Is there a big problem leaving it here right now?

jodh-intel replied:
Maybe a comment stating that this is purely for bootstrapping purposes might swing it ;)

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data:
```diff
@@ -0,0 +1,54 @@
+# Install docker
+# The sourceing of environment is a workaround to get any proxy settings we need
+ - 'bash -c "set -a; source /etc/environment; set +a; curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -"'
+ - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
```
Hard-coded arch.

grahamwhaley replied:
I think I'm going to say 'we are very happy for PRs to update this'... I'll have a look to see if there is a trivial fix, otherwise I think it will stay for the minute

jodh-intel replied:
ok. Now we have three architectures, we really need to try to make all new code architecture agnostic or tolerant atleast. However, if there is urgency to land this or we just aren't sure how best to resolve it for the other arches, how about raising an issue so we don't forget to update this (you could also /cc some folks maybe so they are aware in advance that this is limited to a single arch at this stage).

grahamwhaley replied:
I've added #35, and also added a check so we die at the moment if we are not on x86.

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+      - PUBLICKEYCONTENTS
+
+packages:
+ - gawk
```
You could sort this one too.

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+    path: /home/jenkins/.gitconfig
+# Just can't get these ownerships to work with cloud-init
+# we'll fix them in the runcmd section.
+#    owner: jenkins:jenkins
```
I appreciate this is commented out but shouldn't it be `# USERNAME:USERNAME`?

grahamwhaley replied:
It should - as should a few other places - I'll have a sweep

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+-   content: |
+        [Service]
+        Environment="HTTP_PROXY=http://PROXY_IP:PROXY_PORT"
+        Environment="HTTPS_PROXY=http://PROXY_IP:PROXY_PORT"
```
Shouldn't the value be `https` here?

grahamwhaley replied:
That is site dependent. On the site I tested this at, no, it is not. I'll tweak it to be a bit more regexp though

jodh-intel replied:
Is it normal to have an https proxy variable pointing at an http proxy then?

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+        [http]
+            proxy = http://PROXY_IP:PROXY_PORT
+        [https]
+            proxy = http://PROXY_IP:PROXY_PORT
```
Shouldn't this value be `https`?

grahamwhaley replied:
I'll flip the http's to https's for the https proxy lines - I think (well, know) it does not have to be - I think if you use an http tunnel (I did try to look up the specifics, but proxy stuff quickly gets complex :-), then that is what can make them just http.... it's very site specific anyway, so ultimately the end user has to know what to edit to make it work for them.

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+            proxy = http://PROXY_IP:PROXY_PORT
+        [https]
+            proxy = http://PROXY_IP:PROXY_PORT
+    path: /home/jenkins/.gitconfig
```
`/home/USERNAME/.gitconfig`?

jodh-intel commented on VMs/metrics/ubuntu16.04/user-data.proxy:
```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2018 Intel Corporation
+# 
+# SPDX-License-Identifier: Apache-2.0
+groups:
```
This file has similar content to `user-data`. Is there any way to make cloud-init include the common bits to avoid the duplication?

grahamwhaley replied:
Sadly, not a trivial or even medium-hard way I could find. I started off with them split and `cat`ing them together, and that does not work as `cloud-init` (for some sections?) takes the last seen section - it does not auto-merge them. There are some wacky merge-rules you can set, but I think they do not apply in this instance. So, I looked for YAML merge tools - and that looked quite non-trivial as well - so I went with the split sadly. Yeah, I wasn't too impressed either.

jodh-intel replied:
I guess a follow-on PR could "build" these two yaml files using some `m4`/`sed` magic maybe so we guarantee they are consistent when used?

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+
+### 5_login_clone.sh config_dir
+
+This script is mainly for testing. It will run a VM clone (which should have been created with `4_clone_vm.sh`, and ssh into it.
```
Missing close bracket before comma.

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+details on how to integrate this script into the Jenkins config.
+
+> **Note:** This script is written to try and avoid generating or consuming data from the
+>
```
Extra blank line breaks up the note in a strange way.

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+Jenkins UI should perform a VM shutdown/clone/run cycle, and thus get us back to having
+a stable clean VM on that node.
+
+Below are examples of the configuraiton dialogs:
```
Typo: "configuration".

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+On your slave VM node host machine, create a pair of scripts specific to your host to allow
+Jenkins to invoke the sub-scripts. For example:
+
+start_gitvm.sh
```
Might be worth putting this script name (and the one below) in backticks to make it stand out as a command.

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+# Kata Containers metrics Jenkins CI VMs
```
Could you add a table of contents to this doc?

jodh-intel commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+This directory contains a set of configuration files and helper scripts
+to aid in setting up a [Jenkins](https://jenkins.io/) CI system to run
+the [Kata Containers metrics tests](https://github.com/kata-containers/tests/tree/master/metrics)
+inside Virtual Machines (VMs).
```
I wonder if it's worth adding a link to http://kata-jenkins-ci.westus2.cloudapp.azure.com?

grahamwhaley replied:
Hmm. Maybe, but that is not where these VM images run - that uses, afaik, the Azure base images with some `cloud-init` on top that we store [in our Jenkins xml configs](https://github.com/kata-containers/ci/blob/master/jenkins/config.xml#L192).
Where you thinking of a new section or something to discuss the Azure setup vs the bare-metal setup?

jodh-intel replied:
I'm easy. This was just an idea after I noticed that none of the doc in this CI mentions the location of where you can view the the CI's built from this repo.

jodh-intel commented:
Hi @klynnrif - ptal at the doc content on this PR.

grahamwhaley commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,242 @@
+reduce the 'dirty node' instabilities we saw with pure bare-metal builds.
+
+Also, in the event of getting a dirty VM node, a simple 'node offline/online' in the
+Jenkins UI should perform a VM shutdown/clone/run cycle, and thus get us back to having
```
I think that would change the meaning - please check new version (to be pushed soon)

chavafg commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+
+	sed -i "s/PUBLICKEYCONTENTS/${ESCAPEDPUBLICKEYCONTENTS}/" ./${USER_DATA_NAME}
+	sed -i "s/USERNAME/${VM_USERNAME}/" ./${USER_DATA_NAME}
+
```
remove extra-line?

chavafg commented on VMs/metrics/2_build_baseimage.sh:
```diff
@@ -0,0 +1,144 @@
+run_cloud_init() {
+	message "Running cloud init VM boot"
+
+# first, clean up any pre-existing domain
```
Indent comment

chavafg commented on VMs/metrics/6_start_clone_agent.sh:
```diff
@@ -0,0 +1,76 @@
+if [ $# != 1 ]; then
+	error "Missing first argument"
+	message "Usage: $0 distro_config_dirname"
+	die
```
This would print a `Fatal:` without any other message?  this may create confusion

grahamwhaley replied:
I've found a few more dangling 'die's as well - will have a sweep

grahamwhaley commented on VMs/metrics/1_create_keys.sh:
```diff
@@ -0,0 +1,9 @@
+
+set -x
+
+rm -f virsh.rsa virsh.rsa.pub
```
Good call - the vars already existed (in fact, they had a duplicate) - cleaned up!
Good calls on the other bits as well :-)

grahamwhaley commented:
Updated. Many many fixes:
- README.md updates
- many script fixes
- user-data fixes to make the build actually work again

klynnrif commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,243 @@
+### Proxies
+
+To configure proxies in the cloud-init `user-data` file is complex, as we need
+to set up the proxies for a number of programs, including, `apt`, `git` and `docker`. To
```
Adding in the series comma after "git": to set up the proxies for a number of programs, including, `apt`, `git`, and `docker`.

grahamwhaley replied:
done

grahamwhaley commented on VMs/metrics/README.md:
```diff
@@ -0,0 +1,243 @@
+### 6_start_clone_agent.sh config_dir
+
+This script is designed to be launched by Jenkins. It clones and runs a VM from the
+master image, copy the Jenkins `agent.jar` file across to the image, and execute the
```
done (I suspect a previous edit changed the tone)

klynnrif replied:
@grahamwhaley Thanks! Yes, I suspected so.

grahamwhaley commented:
Updated - added the `sed` replace for `ARCH` - may now work on ARM and ppc64? (I don't know if you guys need to use these scripts or not yet with your CIs, but they are here if you do, and the cloud-init files might make a reference for what needs to be installed etc.).
Also updated the `user-env` files - had to add the git email configs for the user for the PR builds, otherwise some of their funky branch rebases bombed on git (it required the user config to write the commits I guess).

jodh-intel commented:
lgtm
